### PR TITLE
feat(types): expose public interfaces for props and slots

### PR DESCRIPTION
**Pull request by jacekkarczmarczyk**
## Description
Extracts interfaces and types exports to separate files

## Motivation and Context
Vuetify 2.1.11 introduced a bug/breaking change that makes not possible to import simple types/interfaces that are exported from components, for example

`import { DataOptions } from 'vuetify/src/components/VData/VData';`

results in the following error:
```
/node_modules/vuetify/src/util/mergeData.ts(16,19):
16:19 Cannot find name 'Dictionary'.
    14 |
    15 | function parseStyle (style: string) {
  
16 |   const styleMap: Dictionary<any> = {}
       |                   ^
    17 |
    18 |   for (const s of style.split(pattern.styleList)) {
    19 |     let [key, val] = s.split(pattern.styleProp)
```

That's because of importing from `vuetify/src` which is not working very well, however we don't need to import full component file, just the interface definition

## How Has This Been Tested?
`yarn test`

## Other comments
Moving exports from one file to another is technically a breaking change, adding back exports to previous locations (leaves those in new locations) would techncally be a new feature and a quite mess at the same time. 
Could it be treated as a bug fix?

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)
- [ ] I've added new examples to the kitchen (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/util/helpers.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ import Vue, { VNodeData } from 'vue'
 import { VNode, VNodeDirective, FunctionalComponentOptions } from 'vue/types'
 import { VuetifyIcon } from 'vuetify/types/services/icons'
 import mergeData from './mergeData'
+import { compareFn } from '@components/VData/types'
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/util/helpers.ts]: @

File: packages/vuetify/src/components/VCalendar/types.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+export interface VTimestamp {
```
Comment (KaelWD) [inline packages/vuetify/src/components/VCalendar/types.ts]: `d.ts`

File: packages/vuetify/src/components/VDataIterator/VDataIterator.ts [hunk 3]
```diff
@@ -9,10 +9,12 @@ import Themeable from '../../mixins/themeable'
 
 // Helpers
 import { deepEqual, getObjectValueByPath, getPrefixedScopedSlots, getSlot, camelizeObjectKeys } from '../../util/helpers'
-import { DataProps } from '../VData/VData'
-import { PropValidator } from 'vue/types/options'
 import { breaking, removed } from '../../util/console'
 
+// Types
+import { PropValidator } from 'vue/types/options'
+import { DataProps } from '@components/VData/types'
```
Comment (KaelWD) [inline packages/vuetify/src/components/VDataIterator/VDataIterator.ts]: `@`

File: packages/vuetify/src/components/VDataTable/VDataTable.ts [hunk 4]
```diff
@@ -22,8 +20,10 @@ import MobileRow from './MobileRow'
 import ripple from '../../directives/ripple'
 
 // Helpers
-import { deepEqual, getObjectValueByPath, compareFn, getPrefixedScopedSlots, getSlot, defaultFilter, FilterFn, camelizeObjectKeys } from '../../util/helpers'
+import { deepEqual, getObjectValueByPath, getPrefixedScopedSlots, getSlot, defaultFilter, camelizeObjectKeys } from '../../util/helpers'
 import { breaking } from '../../util/console'
+import { TableHeader, FilterFn } from './types'
+import { DataProps, DataOptions, DataPagination, compareFn } from '@components/VData/types'
```
Comment (KaelWD) [inline packages/vuetify/src/components/VDataTable/VDataTable.ts]: `@`

**Comment by johnleider** [comment]
Requesting final comments @KaelWD

**Comment by jacekkarczmarczyk** [comment]
Based on Neko's comment (about this PR being a non-breaking feature) I'm planning to move it to dev and also add few more classes

File: packages/vuetify/src/components/VWindow/VWindow.ts [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
 import './VWindow.sass'
 
-// Components
-import VBtn from '../VBtn'
-import VIcon from '../VIcon'
-import { BaseItemGroup } from '../VItemGroup/VItemGroup'
+// Types
+import { VNode, VNodeDirective } from 'vue/types/vnode'
+import { PropType } from 'vue/types/options'
+import { TouchHandlers } from '@directives/touch/types'
```
Comment (KaelWD) [inline packages/vuetify/src/components/VWindow/VWindow.ts]: @

File: packages/vuetify/src/components/VAutocomplete/VAutocomplete.ts [hunk 6]
```diff
@@ -44,9 +44,9 @@ export default VSelect.extend({
     },
     noFilter: Boolean,
     searchInput: {
-      type: String as PropType<string | undefined>,
+      type: String,
       default: undefined,
-    },
+    } as PropValidator<string | undefined>,
```
Comment (KaelWD) [inline packages/vuetify/src/components/VAutocomplete/VAutocomplete.ts]: Why? We've been slowly moving to PropType since it was added.

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VAutocomplete/VAutocomplete.ts]: Ah ok, I had some problem when I used PropType, so I've changed to PropValidator in few places, but it appeared that the problem was unrelated to it. 
I'll change all PropValidator to PropType then

**Comment by KaelWD** [comment]
Should these all be available in one place (`import { VTimestamp } from 'vuetify'`) or do I have to do `import { VTimestamp } from 'vuetify/src/components/VCalendar/types'`

File: packages/vuetify/src/components/VColorPicker/VColorPickerEdit.ts [hunk 7]
```diff
@@ -10,7 +10,7 @@ import { parseHex } from '../../util/colorUtils'
 
 // Types
 import Vue, { VNode } from 'vue'
-import { PropValidator } from 'vue/types/options'
+import { PropType } from 'vue/types/options'
```
Comment (KaelWD) [inline packages/vuetify/src/components/VColorPicker/VColorPickerEdit.ts]: PropType can be from vue

**Comment by jacekkarczmarczyk** [comment]
```
FilterTreeItemFunction
compareFn
SortItemsFn
DatePickerFormatter
TouchHandlers
VuetifyRuleValidator
VTimestampFormatter
TableHeader
...
```

Names are quite incosistent, maybe we should think about some naming conventions for public types? My proposal for types made public in this PR would be:

| Old name | New name |
| --- | --- |
| FilterTreeItemFunction | VTreeviewItemFunction |
| ItemKey | VSelectItemKey |
| DataOptions | VDataOptions |
| DataPagination | VDataPagination |
| DataProps | VDataScopeProps |
| compareFn | VDataTableCompareFunction |
| SortItemsFn | VDataSortFunction |
| DatePickerFormatter | VDatePickerFormatter |
| AllowedDateFunction | VDatePickerAllowedDatesFunction |
| DateEventColorValue | VDatePickerEventColorValue |
| DateEvents | VDatePickerEvents |
| DateEventColors | VDatePickerEventColors |
| DatePickerType | VDatePickerType |
| DatePickerMultipleFormatter | VDatePickerMultipleFormatter |
| TouchHandlers | VTouchHandlers |
| TouchWrapper | VTouchWrapper |
| TouchValue | VTouchValue |
| VuetifyRuleValidator | VInputValidationRule |
| VuetifyMessage | VInputMessage |
| VuetifyRuleValidations | VInputValidationRules |
| VTimestamp | VCalendarTimestamp |
| VTimestampFormatter | VCalendarFormatter |
| VEventInput | VCalendarEvent |
| FilterFn | VDataTableFilterFunction |
| TableHeader | VDataTableHeader |
| ItemsPerPageOption | VDataItemsPerPageOption |


Thoughts @vuetifyjs/core-team ?

**Comment by KaelWD** [comment]
Lose the prefixes imo, it's all namespaced so there's no chance of a collision.

**Comment by jacekkarczmarczyk** [comment]
You mean `V`? I think it's good to know that it's Vuetify interface, also collisions are possible imho, I can imagine someone having `DataOptions` or `InputMessage` type which not being directly related to Vuetify

**Comment by KaelWD** [comment]
You can `import { x as y }`. Nobody else prefixes types - eg. `Component`, `PropType`, `PluginFunction` from vue.

File: packages/vuetify/src/components/VData/VData.ts [hunk 8]
```diff
@@ -90,7 +63,7 @@ export default Vue.extend({
     disableFiltering: Boolean,
     search: String,
     customFilter: {
-      type: Function as any as PropType<typeof searchItems>,
+      type: Function as PropType<typeof searchItems>,
```
Comment (KaelWD) [inline packages/vuetify/src/components/VData/VData.ts]: sort gets an interface but filter and group don't?

**Comment by KaelWD** [comment]
Next step after this is merged would be to add interface names to the api table in the docs.

**Pull request merged**